[NEW] Add activity indicators for Uploading and Recording using new API; Support thread context; Deprecate the old typing API

sumukhah opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->
# New statuses for user actions/events (Typing 2.0)
Rocket chat sponsored GSoC Project
Mentor: Rodrigo Nascimento
student: Sumukha Hegde

## Primary tasks of this project:
* To implement **Recording**, and **uploading** activity indicators
* To implement thread statuses

### Implementation of Recording and Uploading activity indicators
Showing activity indicators (such as typing, recording, uploading) improves overall user experience and offers a live chat experience to users who may expect quick response times. For this reason, many chat applications have used these activity indicators. 
I have implemented recording and uploading indicators in mobile and in the web application(live chat remaining). We have used a separate activity type (named **user-activity**) that would handle all three activity indicators. We would pass the (typing, recording, uploading) action types as `actionType` parameter while calling the streamer to broadcast these user events.
But the new activity type (**user-activity**) supports only on the web and mobile platform and does not support live chat, bot, e.t.c,. The backend fires the new event when receives the old one and fires the old event when receives the new one, firing both events allows old clients to stay compatible until the next breaking change.

### Implementation of thread statuses
In the previous app, Thread activities were not displaying on the thread but on the main channel. This issue is fixed in this Project, we used a parameter `extra` that would take thread id (tmid) as a value if the user is performing on a thread. Both the mobile and web applications can show these improvements.

### Formats
Old events
```javascript
// Sending event
{
  "msg": "method",
  "id": "1",
  "method": "stream-notify-room",
  "params": [
    "GENERAL/typing",
    "username",
    true, // true when typing, false when not typing
    {
      "token": "cu4B8Aab4ZbowqiwX" // When livechat
    }
  ]
}

// Receiving event
{
  "msg": "changed",
  "collection": "stream-notify-room",
  "id": "id",
  "fields": {
    "eventName": "GENERAL/typing",
    "args": [
      "username",
      true,
      {
        "token": "defaYPat68LevQpLm" // When livechat
      }
    ]
  }
}
```

New events
```javascript
// Sendind event
{
  "msg": "method",
  "id": "1",
  "method": "stream-notify-room",
  "params": [
    "GENERAL/user-activity",
    "username",
    ["user-typing"], // Can be multiple events ["user-typing", "user-recording", "user-uploading"]. Remove the event to stop, pass empty array to stop all
    {
      "tmid": "cu4B8Aab4ZbowqiwX", // When thread
      "token": "defaYPat68LevQpLm" // When livechat
    }
  ]
}

// Receiving event
{
  "msg": "changed",
  "collection": "stream-notify-room",
  "id": "id",
  "fields": {
    "eventName": "GENERAL/user-activity",
    "args": [
      "username",
      ["user-typing"], // Can be single, multiple or empty
      {
        "tmid": "cu4B8Aab4ZbowqiwX", // When thread
        "token": "defaYPat68LevQpLm" // When livechat
      }
    ]
  }
}
```
### References
[#22392 web app PR](https://github.com/RocketChat/Rocket.Chat/pull/22392/files)
[#3243 mobile PR](https://github.com/RocketChat/Rocket.Chat.ReactNative/pull/3243)
[#139 SDK support for mobile](https://github.com/RocketChat/Rocket.Chat.js.SDK/pull/139)

<img src="https://user-images.githubusercontent.com/23723464/129456739-3bd10fa6-baaf-4241-9935-2346adf67a76.png" width="500px">

### extras
https://user-images.githubusercontent.com/23723464/121769591-0a10eb00-cb82-11eb-96d8-241486dbccd4.mov
https://user-images.githubusercontent.com/23723464/122437728-9aeb2a80-cfb7-11eb-945a-dd9131f9df74.mov

rodrigok commented on app/ui/client/lib/UserAction.js:
```diff
@@ -0,0 +1,177 @@
+	return user.username;
+};
+
+const stopActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), false, activityType, extras);
+const startActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), true, activityType, extras);
```
Here you are sending an extra parameter, while the backend is dealing and expecting 3 `username, _activity, extraData` you are passing 4 `shownName(Meteor.user()), false, activityType, extras`.

My suggestion is to merge the boolean and activity type as `shownName(Meteor.user()), {type: activityType, activity: false}, extras`

sumukhah replied:
true, but it won't give an error right now. We can do it while integrating live chat? I didn't touch it because live chat is using it.

rodrigok commented on app/ui/client/lib/UserAction.js:
```diff
@@ -0,0 +1,177 @@
+		}
+		selfActivity[activityType] = true;
+
+		renews[rid] = setTimeout(() => {
```
You can have single `renews` and `timeouts` here, just use `rid + activityType` as key instead of only `rid`. It will make the code simpler.

rodrigok commented on app/ui/client/lib/UserAction.js:
```diff
@@ -0,0 +1,177 @@
+
+	start(rid, activityType, extras = {}) {
+		if (activityType === USER_RECORDING) {
+			this.startPerformingAction(rid, USER_RECORDING, recordingTimeouts, recordingRenews, extras);
```
If you use a sinlge renew and timeout object you can eliminate these ifs and pass the activity type directly

rodrigok commented on app/ui/client/lib/UserAction.js:
```diff
@@ -0,0 +1,177 @@
+	}
+
+	stop(rid, activityType, extras) {
+		if (activityType === USER_RECORDING) {
```
If you use a sinlge renew and timeout object you can eliminate these ifs and pass the activity type directly

rodrigok commented on app/ui/client/lib/UserAction.js:
```diff
@@ -0,0 +1,177 @@
+
+function handleStreamAction(activity, performingUsers, rid, username, extras) {
+	const id = extras?.tmid ? extras.tmid : rid;
+	const users = performingUsers.get(id) || {};
```
Here you can have a global `performingUsers` variable instead of 3 different dicts. And value could have all activities. Eg:

```
const activities = performingUsers.get(id) || {};
const users = activities[activityType];
```

rodrigok commented on app/ui/client/lib/UserAction.js:
```diff
@@ -0,0 +1,177 @@
+	}
+
+	get(roomId, activityType) {
+		if (activityType === USER_RECORDING) {
```
If you unify the dicts you can make this simpler as well, like:
```
const activities = performingUsers.get(roomId) || {};
return Object.keys(activities[activityType] || {})
```

rodrigok commented on app/ui/client/lib/fileUpload.js:
```diff
@@ -55,6 +55,14 @@ export const uploadFileWithMessage = async (rid, tmid, { description, fileName,
 		},
 	});
+	if (Session.get('uploading').length) {
+		UserAction.start(rid, 'user-uploading', { tmid });
```
Should use the exported variable here instead of `user-uploading`

rodrigok commented on app/ui/client/lib/fileUpload.js:
```diff
@@ -55,6 +55,14 @@ export const uploadFileWithMessage = async (rid, tmid, { description, fileName,
+	}
+	const msgUploadingInterval = setInterval(function() {
+		if (Session.get('uploading').length) {
+			UserAction.start(rid, 'user-uploading', { tmid });
```
Same here

rodrigok commented on app/ui/client/lib/fileUpload.js:
```diff
@@ -73,13 +81,23 @@ export const uploadFileWithMessage = async (rid, tmid, { description, fileName,
+		const remainingUploads = Session.set('uploading', uploads.filter((u) => u.id !== upload.id));
+
+		if (!Session.get('uploading').length) {
+			UserAction.stop(rid, 'user-uploading', { tmid });
```
Same here

rodrigok commented on app/ui/client/lib/fileUpload.js:
```diff
@@ -73,13 +81,23 @@ export const uploadFileWithMessage = async (rid, tmid, { description, fileName,
 			u.percentage = 0;
 		});
+		if (!uploads.length) {
+			UserAction.stop(rid, 'user-uploading', { tmid });
```
Same here

rodrigok commented on app/ui-vrecord/client/vrecord.js:
```diff
@@ -32,27 +32,41 @@ Template.vrecDialog.helpers({
+			stopVideoRecording(rid, tmid);
 		} else {
 			VideoRecorder.record();
+			UserAction.start(rid, USER_RECORDING, { tmid });
```
May be better to have an option on `start` to not have a timeout, or pass here a custom timeout, so you don't need to control here the interval to keep posting the event since it already exists in UserAction as `renews`

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,49 @@
+};
+
+Template.userActionIndicator.helpers({
+	typingUsersData() {
```
I believe you can simplify this and return the final strings here instead of parameters to let the template create the string. So the template just mount the `<strong>names</string>activity[, <strong>names</string>activity]`

Example of final string that would be shown: 
**rodrigo, rafael and gabriel** are typing, **marcelo** is recording...

Example of data to be returned here to the template:
```
[{
  users: 'rodrigo, rafael and gabriel',
  activity: 'are typing'
}, {
  users: 'marcelo',
  activity: 'is recording'
}]
```

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.html:
```diff
@@ -0,0 +1,7 @@
+<template name="userActionIndicator" args="rid, tmid">
+	<div class="rc-message-box__activity-indicator">
+		{{> notifyUserActivity data=typingUsersData}}
```
If you accept my suggestion to simplify you can eliminate this file since the render will be dynamic in a look on the other template

rodrigok commented on app/theme/client/imports/components/message-box.css:
```diff
@@ -51,13 +52,14 @@
+	&__activity-indicator {
+		width: 100%;
+		display: grid;
+		grid-template-columns: repeat(auto-fill, minmax(200px, 2fr));
```
Let's use a comma to separate activities

rodrigok commented on app/ui-message/client/messageBox/notifyUserActivity.html:
```diff
@@ -0,0 +1,20 @@
+		<div class="rc-message-box__typing">
+			<span class="rc-message-box__typing-user">{{users}}</span>
+			{{#if multi}}
+				{{#if selfActivity}}
+					{{_ "are_also"}} {{_ action}}
```
We can eliminate this case `also`. Let's remove this control of the current user also typing, not useful and we do not support this on mobiles anyways.

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,49 @@
+	return {
+		action,
+		multi: true,
+		selfActivity: UserAction.selfActivity[activity],
```
We can eliminate this case `also`. Let's remove this control of the current user also typing, not useful and we do not support this on mobiles anyways.

rodrigok commented on app/ui-message/client/messageBox/messageBoxAudioMessage.js:
```diff
@@ -95,6 +119,9 @@ Template.messageBoxAudioMessage.events({
 			}, 1000));
 			recordingRoomId.set(this.rid);
+			recordingIndicatorInterval.set(setInterval(() => {
+				UserAction.start(this.rid, USER_RECORDING, { tmid: this.tmid });
```
Same case I've mentioned for video recording here, if there is an option to start the activity and ignore the timeout this interval is not necessary anymore.

rodrigok commented on app/ui-message/client/messageBox/messageBoxAudioMessage.js:
```diff
@@ -2,18 +2,43 @@ import { ReactiveVar } from 'meteor/reactive-var';
+
+const stopRecording = (rid, tmid) => {
+	const result = new Promise((resolve) => AudioRecorder.stop(resolve));
+	result.then(() => {
```
Please convert to async function and use await here

rodrigok commented on app/ui-message/client/messageBox/messageBoxAudioMessage.js:
```diff
@@ -2,18 +2,43 @@ import { ReactiveVar } from 'meteor/reactive-var';
+const startRecording = (rid, tmid) => {
+	const result = new Promise((resolve, reject) =>
+		AudioRecorder.start((result) => (result ? resolve() : reject())));
+	result.then(() => {
```
Convert to async function and use await here

rodrigok reviewed (changes requested):
Can you convert the new files to typescript?

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,53 @@
+		}
+		const userActions = [];
+		const activities = Object.keys(roomAction) || {};
+		activities.forEach((key) => {
```
Using a map you make the code easier to understand
```suggestion
		const userActions = activities.map((key) => {
```

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,53 @@
+		if (_.isEmpty(roomAction)) {
+			return [];
+		}
+		const userActions = [];
```
```suggestion

```

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,53 @@
+				return;
+			}
+			if (users.length === 1) {
+				userActions.push({
```
```suggestion
				return {
```

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,53 @@
+					multi: false,
+					users: users[0],
+				});
+				return;
```
```suggestion

```

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,53 @@
+			}
+			let usernames = users.slice(0, maxUsernames - 1).join(', ');
+			usernames = [usernames, last];
+			userActions.push({
```
```suggestion
		return {
```

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,53 @@
+				multi: true,
+				users: usernames.join(` ${ t('and') } `),
+			});
+		});
```
And use a filter to remove undefined items
```suggestion
		}).filter((i) => i);
```

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,53 @@
+				users: usernames.join(` ${ t('and') } `),
+			});
+		});
+		if (_.isEmpty(userActions)) {
+			return [];
+		}
```
I don't see the necessity of this return

sumukhah replied:
There is possibility of an empty array. Hence it may give error while setting 		userActions[userActions.length - 1].end = true;

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.html:
```diff
@@ -0,0 +1,17 @@
+	<div class="rc-message-box__action">
+		{{#each data}}
+			<div class="rc-message-box__action-indicator">
+				<span class="rc-message-box__action-user">{{ users}}</span>
```
```suggestion
				<span class="rc-message-box__action-user">{{ users }}</span>
```

rodrigok commented on app/ui/client/lib/UserAction.js:
```diff
@@ -0,0 +1,162 @@
+export const USER_ACTIVITY = 'user-activity';
+
+
+const activityTimeouts = {};
+const activityRenews = {};
```
Use Map for those structures

rodrigok replied:
This was not resolved. My suggestion was for you to use for example `const activityRenews = new Map();`

sumukhah replied:
okay.. But what is the idea behind it? Why would you prefer a map over an object here?

sumukhah commented:
> Can you convert the new files to typescript?

Which all the files? UserAction.js?

rodrigok commented on app/lib/server/startup/settings.js:
```diff
@@ -1022,6 +1022,13 @@ settings.addGroup('General', function() {
 		});
 	});
+	this.section('User_Action_Indicators', function() {
+		this.add('Use_New_Action_Indicator', true, {
```
I'd use the other way around here. Something like `Fire_Old_Typing_Event` and would mention it's a compatibility feature in the description

rodrigok replied:
Of course it would keep firing the new event but would fire the old as well

sumukhah replied:
We will fire only the old typing action (if the option is allowed). Handle function is capable of handling them both. So no need to fire twice.

rodrigok commented on app/ui-message/client/messageBox/userActionIndicator.js:
```diff
@@ -0,0 +1,51 @@
+				users: usernames.join(` ${ t('and') } `),
+			};
+		}).filter((i) => i);
+		if (_.isEmpty(userActions)) {
```
Try to not use the underscore lib, you can use `!userActions.length` or `!Array.isArray(userActions) || userActions.length === 0` instead

rodrigok commented on app/ui/client/lib/fileUpload.js:
```diff
@@ -55,6 +55,14 @@ export const uploadFileWithMessage = async (rid, tmid, { description, fileName,
+	if (Session.get('uploading').length) {
+		UserAction.start(rid, USER_UPLOADING, { tmid });
+	}
+	const msgUploadingInterval = setInterval(function() {
+		if (Session.get('uploading').length) {
+			UserAction.start(rid, USER_UPLOADING, { tmid });
+		}
+	}, 5000);
```
Do you have plans to change this to a method inside `UserAction`?

sumukhah replied:
Yeah.. that's right..This function can be reusable for uploading and recording features. 
I'll do it shortly

sumukhah commented:
I have converted the UserAction.js file to TS. Other than this I had also converted userActionIndicator to TS. But it was a helper method for the userActionIndicator template, and cannot be imported on another template file. So I did not include the file.
